File: backend/app/seed.py
"""
Script para cargar datos de ejemplo en la base de datos
Ejecutar: python -m app.seed
"""

import sys
import os
from decimal import Decimal

# Agregar el directorio raíz al path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from app.core.database import SessionLocal, engine, Base
from app.core.security import get_password_hash
from app.models.user import User, UserRole
from app.models.category import Category
from app.models.product import Product
from app.models.config import StoreConfig, Sequence
from app.models.inventory import InventoryMovement, MovementType


def seed_database():
    """Carga datos de ejemplo"""
    print("Creando tablas...")
    Base.metadata.create_all(bind=engine)

    db = SessionLocal()

    try:
        # Verificar si ya hay datos
        if db.query(User).first():
            print("La base de datos ya tiene datos. Saltando seed.")
            return

        print("Cargando datos de ejemplo...")

        # ==================== USUARIOS ====================
        print("  Creando usuarios...")

        admin = User(
            email="admin@tienda.com",
            username="admin",
            password_hash=get_password_hash("Admin123!"),
            first_name="Administrador",
            last_name="Sistema",
            role=UserRole.ADMIN
        )
        db.add(admin)

        seller = User(
            email="vendedor@tienda.com",
            username="vendedor",
            password_hash=get_password_hash("Vendedor123!"),
            first_name="Juan",
            last_name="Pérez",
            role=UserRole.SELLER
        )
        db.add(seller)

        db.flush()

        # ==================== CATEGORÍAS ====================
        print("  Creando categorías...")

        categories_data = [
            {"name": "Camisetas", "description": "Camisetas y playeras de diversos estilos"},
            {"name": "Pantalones", "description": "Jeans, pantalones casuales y formales"},
            {"name": "Gorras", "description": "Gorras, cachuchas y sombreros"},
            {"name": "Gafas", "description": "Gafas de sol y accesorios"},
            {"name": "Accesorios", "description": "Cinturones, billeteras y otros accesorios"},
        ]

        categories = {}
        for cat_data in categories_data:
            cat = Category(**cat_data)
            db.add(cat)
            db.flush()
            categories[cat_data["name"]] = cat

        # ==================== PRODUCTOS ====================
        print("  Creando productos...")

        products_data = [
            # Camisetas
            {
                "sku": "CAM-001",
                "barcode": "7501234567890",
                "name": "Camiseta Básica Blanca",
                "description": "Camiseta de algodón 100%, cuello redondo",
                "category": "Camisetas",
                "brand": "BasicWear",
                "size": "M",
                "color": "Blanco",
                "sale_price": Decimal("15990"),
                "cost_price": Decimal("8000"),
                "tax_rate": Decimal("0.19"),
                "min_stock": 10,
                "initial_stock": 50
            },
            {
                "sku": "CAM-002",
                "barcode": "7501234567891",
                "name": "Camiseta Básica Negra",
                "description": "Camiseta de algodón 100%, cuello redondo",
                "category": "Camisetas",
                "brand": "BasicWear",
                "size": "M",
                "color": "Negro",
                "sale_price": Decimal("15990"),
                "cost_price": Decimal("8000"),
                "tax_rate": Decimal("0.19"),
                "min_stock": 10,
                "initial_stock": 45
            },
            {
                "sku": "CAM-003",
                "barcode": "7501234567892",
                "name": "Polo Premium Azul",
                "description": "Polo de algodón piqué con bordado",
                "category": "Camisetas",
                "brand": "PoloStyle",
                "size": "L",
                "color": "Azul",
                "sale_price": Decimal("29990"),
                "cost_price": Decimal("15000"),
                "tax_rate": Decimal("0.19"),
                "min_stock": 5,
                "initial_stock": 30
            },
            # Pantalones
            {
                "sku": "PAN-001",
                "barcode": "7501234567893",
                "name": "Jean Clásico Azul",
                "description": "Jean de mezclilla, corte recto",
                "category": "Pantalones",
                "brand": "DenimCo",
                "size": "32",
                "color": "Azul",
                "sale_price": Decimal("39990"),
                "cost_price": Decimal("20000"),
                "tax_rate": Decimal("0.19"),
                "min_stock": 8,
                "initial_stock": 25
            },
            {
                "sku": "PAN-002",
                "barcode": "7501234567894",
                "name": "Pantalón Casual Beige",
                "description": "Pantalón de gabardina, corte slim",
                "category": "Pantalones",
                "brand": "CasualFit",
                "size": "34",
                "color": "Beige",
                "sale_price": Decimal("34990"),
                "cost_price": Decimal("17500"),
                "tax_rate": Decimal("0.19"),
                "min_stock": 5,
                "initial_stock": 20
            },
            # Gorras
            {
                "sku": "GOR-001",
                "barcode": "7501234567895",
                "name": "Gorra Baseball Negra",
                "description": "Gorra de baseball con ajuste trasero",
                "category": "Gorras",
                "brand": "CapStyle",
                "size": "Única",
                "color": "Negro",
                "sale_price": Decimal("12990"),
                "cost_price": Decimal("6500"),
                "tax_rate": Decimal("0.19"),
                "min_stock": 15,
                "initial_stock": 40
            },
            {
                "sku": "GOR-002",
                "barcode": "7501234567896",
                "name": "Gorra Trucker Roja",
                "description": "Gorra trucker con malla trasera",
                "category": "Gorras",
                "brand": "CapStyle",
                "size": "Única",
                "color": "Rojo",
                "sale_price": Decimal("14990"),
                "cost_price": Decimal("7500"),
                "tax_rate": Decimal("0.19"),
                "min_stock": 10,
                "initial_stock": 35
            },
            # Gafas
            {
                "sku": "GAF-001",
                "barcode": "7501234567897",
                "name": "Gafas de Sol Aviador",
                "description": "Gafas estilo aviador con protección UV",
                "category": "Gafas",
                "brand": "SunVision",
                "size": "Única",
                "color": "Dorado/Negro",
                "sale_price": Decimal("24990"),
                "cost_price": Decimal("12500"),
                "tax_rate": Decimal("0.19"),
                "min_stock": 10,
                "initial_stock": 25
            },
            {
                "sku": "GAF-002",
                "barcode": "7501234567898",
                "name": "Gafas de Sol Wayfarer",
                "description": "Gafas estilo wayfarer clásico",
                "category": "Gafas",
                "brand": "SunVision",
                "size": "Única",
                "color": "Negro",
                "sale_price": Decimal("22990"),
                "cost_price": Decimal("11500"),
                "tax_rate": Decimal("0.19"),
                "min_stock": 10,
                "initial_stock": 30
            },
            # Accesorios
            {
                "sku": "ACC-001",
                "barcode": "7501234567899",
                "name": "Cinturón Cuero Negro",
                "description": "Cinturón de cuero genuino con hebilla metálica",
                "category": "Accesorios",
                "brand": "LeatherCraft",
                "size": "M",
                "color": "Negro",
                "sale_price": Decimal("19990"),
                "cost_price": Decimal("10000"),
                "tax_rate": Decimal("0.19"),
                "min_stock": 8,
                "initial_stock": 20
            },
            {
                "sku": "ACC-002",
                "barcode": "7501234567900",
                "name": "Billetera Cuero Café",
                "description": "Billetera de cuero con múltiples compartimentos",
                "category": "Accesorios",
                "brand": "LeatherCraft",
                "size": "Única",
                "color": "Café",
                "sale_price": Decimal("29990"),
                "cost_price": Decimal("15000"),
                "tax_rate": Decimal("0.19"),
                "min_stock": 5,
                "initial_stock": 15
            },
            {
                "sku": "ACC-003",
                "barcode": "7501234567901",
                "name": "Reloj Deportivo Digital",
                "description": "Reloj digital resistente al agua",
                "category": "Accesorios",
                "brand": "SportTime",
                "size": "Única",
                "color": "Negro",
                "sale_price": Decimal("49990"),
                "cost_price": Decimal("25000"),
                "tax_rate": Decimal("0.19"),
                "min_stock": 3,
                "initial_stock": 10
            },
        ]

        for prod_data in products_data:
            category_name = prod_data.pop("category")
            initial_stock = prod_data.pop("initial_stock")

            product = Product(
                **prod_data,
                category_id=categories[category_name].id,
                current_stock=initial_stock
            )
            db.add(product)
            db.flush()

            # Crear movimiento de inventario inicial
            if initial_stock > 0:
                movement = InventoryMovement(
                    product_id=product.id,
                    user_id=admin.id,
                    type=MovementType.ENTRY,
                    quantity=initial_stock,
                    previous_stock=0,
                    new_stock=initial_stock,
                    reason="Stock inicial"
                )
                db.add(movement)

        # ==================== CONFIGURACIÓN ====================
        print("  Creando configuración de tienda...")

        config = StoreConfig(
            store_name="Mi Tienda de Ropa",
            store_address="Calle Principal 123, Ciudad",
            store_phone="+56 9 1234 5678",
            store_email="contacto@mitienda.com",
            store_rut="12.345.678-9",
            primary_color="#3B82F6",
            secondary_color="#1E40AF",
            accent_color="#F59E0B",
            receipt_header="¡Bienvenido a nuestra tienda!",
            receipt_footer="Gracias por su compra. Vuelva pronto.",
            currency_symbol="$",
            currency_code="CLP"
        )
        db.add(config)

        # ==================== SECUENCIA ====================
        print("  Creando secuencia de recibos...")

        sequence = Sequence(
            name="receipt",
            prefix="R",
            current_value=0,
            padding=8
        )
        db.add(sequence)

        db.commit()
        print("\n✅ Datos de ejemplo cargados exitosamente!")
        print("\nCredenciales de acceso:")
        print("  Admin:    admin@tienda.com / Admin123!")
        print("  Vendedor: vendedor@tienda.com / Vendedor123!")

    except Exception as e:
        db.rollback()
        print(f"\n❌ Error al cargar datos: {e}")
        raise
    finally:
        db.close()


if __name__ == "__main__":
    seed_database()

File: backend/prisma/seed.ts
import { PrismaClient, UserRole } from '@prisma/client';
import bcrypt from 'bcryptjs';

const prisma = new PrismaClient();

async function main() {
  console.log('🌱 Iniciando seed de base de datos...');

  // Crear configuración de tienda
  const storeConfig = await prisma.storeConfig.upsert({
    where: { id: 'store_config' },
    update: {},
    create: {
      id: 'store_config',
      storeName: 'Mi Tienda de Ropa',
      storeNit: '900123456-7',
      storeAddress: 'Calle 123 #45-67, Local 101',
      storePhone: '+57 300 123 4567',
      storeEmail: 'contacto@mitienda.com',
      primaryColor: '#3B82F6',
      secondaryColor: '#1E40AF',
      accentColor: '#F59E0B',
      defaultTaxRate: 0.19, // IVA 19%
      receiptFooter: '¡Gracias por su compra! Visítenos pronto.',
    },
  });
  console.log('✅ Configuración de tienda creada');

  // Crear secuencias
  await prisma.sequence.upsert({
    where: { id: 'receipt_number' },
    update: {},
    create: { id: 'receipt_number', currentValue: 0 },
  });

  await prisma.sequence.upsert({
    where: { id: 'return_number' },
    update: {},
    create: { id: 'return_number', currentValue: 0 },
  });
  console.log('✅ Secuencias creadas');

  // Crear usuario administrador
  const adminPassword = await bcrypt.hash('Admin123!', 12);
  const admin = await prisma.user.upsert({
    where: { email: 'admin@tienda.com' },
    update: {},
    create: {
      email: 'admin@tienda.com',
      username: 'admin',
      passwordHash: adminPassword,
      firstName: 'Administrador',
      lastName: 'Sistema',
      role: UserRole.ADMIN,
    },
  });
  console.log('✅ Usuario administrador creado: admin@tienda.com / Admin123!');

  // Crear usuario vendedor
  const sellerPassword = await bcrypt.hash('Vendedor123!', 12);
  const seller = await prisma.user.upsert({
    where: { email: 'vendedor@tienda.com' },
    update: {},
    create: {
      email: 'vendedor@tienda.com',
      username: 'vendedor',
      passwordHash: sellerPassword,
      firstName: 'Juan',
      lastName: 'Vendedor',
      role: UserRole.SELLER,
    },
  });
  console.log('✅ Usuario vendedor creado: vendedor@tienda.com / Vendedor123!');

  // Crear categorías
  const categories = await Promise.all([
    prisma.category.upsert({
      where: { name: 'Camisetas' },
      update: {},
      create: { name: 'Camisetas', description: 'Camisetas y polos' },
    }),
    prisma.category.upsert({
      where: { name: 'Pantalones' },
      update: {},
      create: { name: 'Pantalones', description: 'Jeans, pantalones casuales y formales' },
    }),
    prisma.category.upsert({
      where: { name: 'Gorras' },
      update: {},
      create: { name: 'Gorras', description: 'Gorras y sombreros' },
    }),
    prisma.category.upsert({
      where: { name: 'Gafas' },
      update: {},
      create: { name: 'Gafas', description: 'Gafas de sol y accesorios' },
    }),
    prisma.category.upsert({
      where: { name: 'Accesorios' },
      update: {},
      create: { name: 'Accesorios', description: 'Cinturones, carteras y otros' },
    }),
  ]);
  console.log('✅ Categorías creadas');

  // Crear productos de ejemplo
  const products = [
    // Camisetas
    {
      sku: 'CAM-001',
      barcode: '7501234567890',
      name: 'Camiseta Básica Blanca',
      description: 'Camiseta 100% algodón',
      categoryId: categories[0].id,
      brand: 'Marca Propia',
      size: 'M',
      color: 'Blanco',
      salePrice: 35000,
      costPrice: 18000,
      taxRate: 0.19,
      minStock: 10,
      currentStock: 50,
    },
    {
      sku: 'CAM-002',
      barcode: '7501234567891',
      name: 'Camiseta Básica Negra',
      description: 'Camiseta 100% algodón',
      categoryId: categories[0].id,
      brand: 'Marca Propia',
      size: 'M',
      color: 'Negro',
      salePrice: 35000,
      costPrice: 18000,
      taxRate: 0.19,
      minStock: 10,
      currentStock: 45,
    },
    {
      sku: 'CAM-003',
      barcode: '7501234567892',
      name: 'Polo Premium Azul',
      description: 'Polo con cuello, tejido premium',
      categoryId: categories[0].id,
      brand: 'Premium Wear',
      size: 'L',
      color: 'Azul',
      salePrice: 75000,
      costPrice: 40000,
      taxRate: 0.19,
      minStock: 5,
      currentStock: 20,
    },
    // Pantalones
    {
      sku: 'PAN-001',
      barcode: '7501234567893',
      name: 'Jean Clásico Azul',
      description: 'Jean corte recto',
      categoryId: categories[1].id,
      brand: 'Denim Co',
      size: '32',
      color: 'Azul',
      salePrice: 89000,
      costPrice: 45000,
      taxRate: 0.19,
      minStock: 8,
      currentStock: 30,
    },
    {
      sku: 'PAN-002',
      barcode: '7501234567894',
      name: 'Pantalón Casual Beige',
      description: 'Pantalón casual de algodón',
      categoryId: categories[1].id,
      brand: 'Casual Style',
      size: '30',
      color: 'Beige',
      salePrice: 65000,
      costPrice: 32000,
      taxRate: 0.19,
      minStock: 5,
      currentStock: 15,
    },
    // Gorras
    {
      sku: 'GOR-001',
      barcode: '7501234567895',
      name: 'Gorra Baseball Negra',
      description: 'Gorra ajustable con visera curva',
      categoryId: categories[2].id,
      brand: 'Sport Cap',
      color: 'Negro',
      salePrice: 28000,
      costPrice: 12000,
      taxRate: 0.19,
      minStock: 15,
      currentStock: 40,
    },
    {
      sku: 'GOR-002',
      barcode: '7501234567896',
      name: 'Gorra Trucker Roja',
      description: 'Gorra estilo trucker con malla',
      categoryId: categories[2].id,
      brand: 'Urban Style',
      color: 'Rojo',
      salePrice: 32000,
      costPrice: 15000,
      taxRate: 0.19,
      minStock: 10,
      currentStock: 25,
    },
    // Gafas
    {
      sku: 'GAF-001',
      barcode: '7501234567897',
      name: 'Gafas de Sol Aviador',
      description: 'Gafas estilo aviador con protección UV',
      categoryId: categories[3].id,
      brand: 'SunVision',
      color: 'Dorado/Negro',
      salePrice: 55000,
      costPrice: 25000,
      taxRate: 0.19,
      minStock: 10,
      currentStock: 35,
    },
    {
      sku: 'GAF-002',
      barcode: '7501234567898',
      name: 'Gafas de Sol Wayfarer',
      description: 'Gafas clásicas estilo wayfarer',
      categoryId: categories[3].id,
      brand: 'SunVision',
      color: 'Negro Mate',
      salePrice: 48000,
      costPrice: 22000,
      taxRate: 0.19,
      minStock: 10,
      currentStock: 28,
    },
    // Accesorios
    {
      sku: 'ACC-001',
      barcode: '7501234567899',
      name: 'Cinturón de Cuero Negro',
      description: 'Cinturón de cuero genuino',
      categoryId: categories[4].id,
      brand: 'Leather Pro',
      size: 'M',
      color: 'Negro',
      salePrice: 42000,
      costPrice: 20000,
      taxRate: 0.19,
      minStock: 8,
      currentStock: 22,
    },
    {
      sku: 'ACC-002',
      barcode: '7501234567800',
      name: 'Billetera Cuero Café',
      description: 'Billetera de cuero con múltiples compartimentos',
      categoryId: categories[4].id,
      brand: 'Leather Pro',
      color: 'Café',
      salePrice: 58000,
      costPrice: 28000,
      taxRate: 0.19,
      minStock: 5,
      currentStock: 18,
    },
    // Producto con bajo stock para pruebas
    {
      sku: 'CAM-004',
      barcode: '7501234567801',
      name: 'Camiseta Edición Especial',
      description: 'Camiseta de edición limitada',
      categoryId: categories[0].id,
      brand: 'Exclusive',
      size: 'S',
      color: 'Multicolor',
      salePrice: 95000,
      costPrice: 50000,
      taxRate: 0.19,
      minStock: 5,
      currentStock: 3, // Bajo stock
    },
  ];

  for (const product of products) {
    await prisma.product.upsert({
      where: { sku: product.sku },
      update: {},
      create: product,
    });
  }
  console.log(`✅ ${products.length} productos creados`);

  // Crear movimientos de inventario iniciales
  const allProducts = await prisma.product.findMany();
  for (const product of allProducts) {
    await prisma.inventoryMovement.create({
      data: {
        productId: product.id,
        userId: admin.id,
        type: 'ENTRY',
        quantity: product.currentStock,
        previousStock: 0,
        newStock: product.currentStock,
        unitCost: product.costPrice,
        reason: 'Stock inicial de apertura',
      },
    });
  }
  console.log('✅ Movimientos de inventario iniciales creados');

  // Registrar en auditoría
  await prisma.auditLog.create({
    data: {
      userId: admin.id,
      action: 'SEED',
      entity: 'system',
      description: 'Base de datos inicializada con datos de ejemplo',
    },
  });

  console.log('');
  console.log('🎉 Seed completado exitosamente!');
  console.log('');
  console.log('📋 Credenciales de acceso:');
  console.log('   Administrador: admin@tienda.com / Admin123!');
  console.log('   Vendedor: vendedor@tienda.com / Vendedor123!');
  console.log('');
}

main()
  .catch((e) => {
    console.error('❌ Error en seed:', e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
